File: LeetCode Practice Problems/103-Binary Tree Zigzag Level Order Traversal.cpp
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        if(root == NULL)
            return ans;
        
        queue<TreeNode*> q;
        q.push(root);
        bool flag = false;
        while(!q.empty()){
            queue<TreeNode*> t;
            vector<int> res;
            while(!q.empty()){
                TreeNode* r = q.front();
                q.pop();
                if(flag){
                    int v = r->val;
                    res.insert(res.begin() , v);
                } else{
                    res.push_back(r->val);
                }
                if(r->left != NULL)
                    t.push(r->left);
                if(r->right != NULL)
                    t.push(r->right);
            }
            ans.push_back(res);
            q = t;
            flag = !flag;
        }
        
        return ans;
    }
};

File: LeetCode/LeetCode Practice Problems/103-Binary Tree Zigzag Level Order Traversal.java
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList();
        Queue<TreeNode> q = new LinkedList();
        if(root != null) {
            q.add(root);
        }
        
        boolean reverse = false;
        while(!q.isEmpty()) {
            List<Integer> res = new ArrayList();
            int sz = q.size();
            for(int i=0;i<sz;i++) {
                TreeNode tmp = q.poll();
                res.add(tmp.val);
                
                if(tmp.left != null) 
                    q.add(tmp.left);
                if(tmp.right != null)
                    q.add(tmp.right);
            }
            if(reverse) {
                Collections.reverse(res);
            }
            ans.add(res);
            reverse = !reverse;
        }
        return ans;
    }
}
